pkg/beholder/client.go

```diff
@@ -0,0 +1,332 @@
+	Meter() otelmetric.Meter
+	Emitter() Emitter
+	Close() error
+}
+
+var _ Client = (*otelClient)(nil)
+
+type messageEmitter struct {
+	exporter      sdklog.Exporter
+	messageLogger otellog.Logger
+}
+
+type otelClient struct {
+	config Config
+	// Logger
+	logger otellog.Logger
+	// Tracer
+	tracer oteltrace.Tracer
+	// Meter
+	meter otelmetric.Meter
+	// Message Emitter
+	emitter Emitter
+	// Graceful shutdown for tracer, meter, logger providers
+	closeFunc func() error
+}
+
+func NewClient(
+	config Config,
+	logger otellog.Logger,
+	tracer oteltrace.Tracer,
+	meter otelmetric.Meter,
+	emitter Emitter,
+	onClose func() error,
+) Client {
+	return &otelClient{
+		config:    config,
+		logger:    logger,
+		tracer:    tracer,
+		meter:     meter,
+		emitter:   emitter,
+		closeFunc: onClose,
+	}
+}
+
+// NewOtelClient creates a new Client with OTel exporter
+func NewOtelClient(cfg Config, errorHandler errorHandlerFunc) (Client, error) {
+	factory := func(ctx context.Context, options ...otlploggrpc.Option) (sdklog.Exporter, error) {
+		return otlploggrpc.New(ctx, options...)
+	}
+	return newOtelClient(cfg, errorHandler, factory)
+}
+
+// Used for testing to override the default exporter
+type otlploggrpcFactory func(ctx context.Context, options ...otlploggrpc.Option) (sdklog.Exporter, error)
+
+func newOtelClient(cfg Config, errorHandler errorHandlerFunc, otlploggrpcNew otlploggrpcFactory) (Client, error) {
+	ctx := context.Background()
+	baseResource, err := newOtelResource(cfg)
+	if err != nil {
+		return nil, err
+	}
+	creds := insecure.NewCredentials()
+	if !cfg.InsecureConnection && cfg.CACertFile != "" {
+		creds, err = credentials.NewClientTLSFromFile(cfg.CACertFile, "")
+		if err != nil {
+			return nil, err
+		}
+	}
+	sharedLogExporter, err := otlploggrpcNew(
+		ctx,
+		otlploggrpc.WithTLSCredentials(creds),
+		otlploggrpc.WithEndpoint(cfg.OtelExporterGRPCEndpoint),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Logger
+	loggerProcessor := sdklog.NewBatchProcessor(
+		sharedLogExporter,
+		sdklog.WithExportTimeout(cfg.LogExportTimeout), // Default is 30s
+	)
+	loggerAttributes := []attribute.KeyValue{
+		attribute.String("beholder_data_type", "zap_log_message"),
+	}
+	loggerResource, err := sdkresource.Merge(
+		sdkresource.NewSchemaless(loggerAttributes...),
+		baseResource,
+	)
+	if err != nil {
+		return nil, err
+	}
+	loggerProvider := sdklog.NewLoggerProvider(
+		sdklog.WithResource(loggerResource),
+		sdklog.WithProcessor(loggerProcessor),
+	)
+	logger := loggerProvider.Logger(cfg.PackageName)
+
+	// Set global logger provider
+	otelglobal.SetLoggerProvider(loggerProvider)
+
+	// Tracer
+	tracerProvider, err := newTracerProvider(cfg, baseResource, creds)
+	if err != nil {
+		return nil, err
+	}
+	tracer := tracerProvider.Tracer(cfg.PackageName)
+	otel.SetTracerProvider(tracerProvider)
+	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}, propagation.Baggage{}))
```

Comment:
Auto-setting globals prevents us from using this in core, which is already setting its own globals. Can we provide a more flexible API here? What if this returned without calling any global funcs? Can we give the caller the option of doing the global calls themselves, or otherwise opting in explicitly for them?

Reply:
IIUC our situation requires a single provider with multiple exporters.

Reply:
Makes sense, will provide a separate way to set globals

Reply:
It looks like we can call `RegisterSpanProcessor` on a global `TracerProvider`, so we don't technically need to constructor the provider with both exporters on hand, but we still need to ultimately have some control of this.

Reply:
@jmank88, addressed [this](https://github.com/smartcontractkit/chainlink-common/pull/694#discussion_r1714389524) comment [here](https://github.com/smartcontractkit/chainlink-common/pull/694/commits/14b67f1738ef1a7c193fcb1367f3d041e61ec336)

Reply:
@jmank88 What is the use case for RegisterSpanProcessor?
I made providers to be exported (see [here](https://github.com/smartcontractkit/chainlink-common/blob/14b67f1738ef1a7c193fcb1367f3d041e61ec336/pkg/beholder/client.go#L53-L55)) so it's possible to call it like this `OtelClient.TracerProvider.RegisterSpanProcessor`.

Reply:
Core and LOOP Plugins both call in to the existing `SetupTracing` helper which creates and registers its own provider: https://github.com/smartcontractkit/chainlink-common/blob/main/pkg/loop/telem.go#L124-L134

IIUC We will have to configure just one provider/resource/set-of-attributes, but the provider can have multiple batchers/exporters/processors.

Reply:
Most of the examples I have seen for OTel logger,tracer,meter have one provider configured per combination of single batcher/exporter/processor. 

You can have multiple providers but only one global for logger,tracer,meter. 
OTel resource/set-of-attributes could and should be shared between providers. 
OTel Resource support is merging so you can have a base resource and extend it if needed with custom attributes per provider.

Reply:
> You can have multiple providers but only one global per logger,tracer,meter.

But AFAIK in our case we do not want two providers because we want to send the same data to both places. The goal is for existing trace usages to be picked up and sent through to beholder as well.

Reply:
> @jmank88, I see that it's used for Loop GRPC endpoint instrumentation via [otelgrpc](https://github.com/smartcontractkit/chainlink-common/blob/a4bce717e58335b066c65ab01f0dd3977215c6c2/pkg/loop/telem.go#L152) library. Is that OTel tracing enabled on production for Loops?

It is opt-in per node op via the `Tracing` section of their config.toml file: https://github.com/smartcontractkit/chainlink/blob/develop/docs/CONFIG.md#tracing

Reply:
> But AFAIK in our case we do not want two providers because we want to send the same data to both places. The goal is for existing trace usages to be picked up and sent through to beholder as well.

yes, that's correct

Reply:
@jmank88, @patrickhuie19  I thought initially that we might want to do that. Now I'm not sure. Let's clarify this.

Reply:
Answered here https://github.com/smartcontractkit/chainlink-common/pull/696#discussion_r1727514619